awscli/customizations/history/show.py

```diff
@@ -0,0 +1,486 @@
+            return self._get_pretty_xml(body)
+        elif self._is_json_structure(body):
+            return self._get_pretty_json(body)
+        else:
+            return body
+
+    def _get_pretty_xml(self, body):
+        # The body is parsed and whitespace is stripped because some services
+        # like ec2 already return pretty XML and if toprettyxml() was applied
+        # to it, it will add even more newlines and spaces on top of it.
+        # So this just removes all whitespace from the start to prevent the
+        # chance of adding to much newlines and spaces when toprettyxml()
+        # is called.
+        stripped_body = self._strip_whitespace(body)
+        xml_dom = xml.dom.minidom.parseString(stripped_body)
+        return xml_dom.toprettyxml(indent=' '*4, newl='\n')
+
+    def _get_pretty_json(self, body):
+        # The json body is loaded so it can be dumped in a format that
+        # is desired.
+        obj = json.loads(body)
+        return self._pformat_dictionary(obj)
+
+    def _is_xml(self, body):
+        try:
+            xml.dom.minidom.parseString(body)
+        except xml.parsers.expat.ExpatError:
+            return False
+        return True
+
+    def _strip_whitespace(self, xml_string):
+        xml_dom = xml.dom.minidom.parseString(xml_string)
+        return ''.join(
+            [line.strip() for line in xml_dom.toxml().splitlines()]
+        )
+
+    def _is_json_structure(self, body):
+        if body.startswith('{'):
+            try:
+                json.loads(body)
+                return True
+            except json.decoder.JSONDecodeError:
+                return False
+        return False
+
+
+class OutputStreamFactory(object):
+    def __init__(self, popen=None):
+        self._popen = popen
+        if popen is None:
+            self._popen = subprocess.Popen
+
+    def get_output_stream(self, stream_type):
+        """Get an output stream to write to
+
+        The value is wrapped in a context manager so make sure to use
+        a with statement.
+
+        :type stream_type: string
+        :param stream_type: The name of the stream to get. Valid values
+             consist of pager and stdout.
+        """
+        if stream_type == 'pager':
+            return self._get_pager_stream()
+        elif stream_type == 'stdout':
+            return self._get_stdout_stream()
+        else:
+            raise ValueError(
+                'Stream type of %s is not supported' % stream_type)
+
+    @contextlib.contextmanager
+    def _get_pager_stream(self):
+        popen_kwargs = self._get_process_pager_kwargs()
+        try:
+            process = self._popen(**popen_kwargs)
+            yield process.stdin
+        finally:
+            process.communicate()
+
+    @contextlib.contextmanager
+    def _get_stdout_stream(self):
+        yield get_binary_stdout()
+
+    def _get_process_pager_kwargs(self):
+        kwargs = get_popen_kwargs_for_pager_cmd(
```

Comment:
I was playing around with this, and one thing I noticed is that we assume that if a user sets a pager via the `PAGER` env var then it can handle color.  I've always had my default `PAGER` set to `less`, so when I first ran the `show` command I had all the escape sequences in my term:

```
$ aws history show
```
![image](https://user-images.githubusercontent.com/368057/33282403-6986e60a-d35d-11e7-92ea-368094c6b140.png)

I just used `--color on | less -R` to get what I wanted (paginated color), but I wonder if there's anything we can do here.  Just a guess, but I'd think `export PAGER=less` is a pretty common configuration.  What do you think?

Reply:
@jamesls I wonder if we should just add -R onto the command if we detect that the value is just `less` with no arguments? @stealthycoin is runs into a similar scenario when using `-S` for the `list` command.

Reply:
Also do you usually get color when you use `less`? I think there may be an environment value `LESS` that sets these flags.

Reply:
> Also do you usually get color when you use less? 

No I don't.

Capturing offline feedback:

* Don't use their `PAGER` var if set.
* Add support for our own env var for paging in the future.